fedimint-core/src/db/mod.rs

```diff
@@ -216,6 +223,80 @@ impl Database {
                 .expect("db autocommit retry counter overflowed");
         } // end of loop
     }
+
+    /// Waits for key to be notified.
+    ///
+    /// Calls the `f` on every notifications. Return the value when `f` returns
+    /// a `Some(T)`
+    pub async fn wait_key_check<'a, K, T>(
+        &'a self,
+        key: &K,
+        f: impl Fn(Option<K::Value>) -> Result<Option<T>>,
+    ) -> Result<(T, DatabaseTransaction<'a>)>
+    where
+        K: DatabaseKey + DatabaseRecord,
+    {
+        // not compile time yet :/
+        assert!(
+            <K as DatabaseKey>::NOTIFY,
+            "wait_key called with key({}) without NOTIFY flag",
+            std::any::type_name::<K>()
+        );
```

Comment:
I wonder if `const _: () = assert!(<K as DatabaseKey>::NOTIFY);` would work ([source stackoverflow](https://stackoverflow.com/a/70925999)).

Alternatively there's always type-level booleans, not that they'd be a good idea typically :laughing: 

```rust
pub trait Bool {
    fn to_bool() -> bool;
}

pub struct True {}
pub struct False {}

impl Bool for True {
    fn to_bool() -> bool {
        true
    }
}
impl Bool for False {
    fn to_bool() -> bool {
        false
    }
}
```